app/src/main/java/net/gsantner/markor/activity/DocumentEditFragment.java

```diff
@@ -551,29 +578,25 @@ public void onFsViewerConfig(FilesystemViewerData.Options dopt) {
         }
     }
 
-    private long _lastChangedThreadStart = 0;
-
     @OnTextChanged(value = R.id.document__fragment__edit__highlighting_editor, callback = OnTextChanged.Callback.TEXT_CHANGED)
     public void onContentEditValueChanged(CharSequence text) {
-        if ((_lastChangedThreadStart + HISTORY_DELTA) < System.currentTimeMillis()) {
-            _lastChangedThreadStart = System.currentTimeMillis();
-            _hlEditor.postDelayed(() -> {
-                Activity activity = getActivity();
-                if (activity instanceof AppCompatActivity) {
-                    ((AppCompatActivity) activity).supportInvalidateOptionsMenu();
-                }
-            }, HISTORY_DELTA);
-        }
-        Activity activity = getActivity();
-        if (activity != null && activity instanceof AppCompatActivity) {
-            ((AppCompatActivity) activity).supportInvalidateOptionsMenu();
-        }
+        checkTextChangeState();
+        updateUndoRedoIconStates();
+    }
 
+    public void checkTextChangeState() {
+        if (_isTextChanged == _document.isContentSame(_hlEditor.getText())) {
+            _isTextChanged = !_isTextChanged;
+
+            if (_saveMenuItem != null) {
+                _saveMenuItem.setEnabled(_isTextChanged).getIcon().mutate().setAlpha(_isTextChanged ? 255 : 40);
```

Comment:
New approach - just enable / disable save icon and grey it out to show save state. Much cleaner than adding a *

Reply:
I personally prefer this way of showing. Not sure it user is happy with it. But less bloat = better, hope they can agree.